nucypher/utilities/concurrency.py

```diff
@@ -337,6 +311,27 @@ def _produce_values(self):
 
         self._result_queue.put(PRODUCER_STOPPED)
 
+    def _format_failures(self, result) -> str:
+        """
+        Performs some basic formatting of the WorkerPool failures,
+        providing some context of why TimedOut/OutOfValues occurred.
+        """
+        failures = self.get_failures()
+        failure_result = ""
+        if result == TIMEOUT_TRIGGERED:
+            failure_result = f"Execution timed out after {self._timeout}s"
+        elif result == PRODUCER_STOPPED:
+            failure_result = f"Execution stopped before completion - not enough available values"
+
+        if failures:
+            # Using one random failure
+            # Most probably they're all the same anyway.
+            value = list(failures)[0]
+            _exception_cls, exception, tb = failures[value]
+            return f"{failure_result} ({len(failures)} failures recorded); for example, {exception}"
```

Comment:
Why did you get rid of printing the traceback? I think it was pretty useful.

Reply:
The traceback was excessive for an exception string message. If necessary, the failures can be obtained from the WorkerPool directly if needed for logging - I guess one more direct alternative to this is that we can have the `OutOfValues` and `TimedOut` exceptions store the copy of the failures dict, so the failures can be directly accessed by anything that catches those exceptions and cares about the underlying failures. This way there will still be access to the failures...I think I'll add this in.

Reply:
@fjarri See https://github.com/nucypher/nucypher/pull/2772/commits/5bde88630e4c98eb36d9b318ac9051edb93c1be9

Reply:
The thing that worries me about this approach is that these traceback are a very useful debugging tool. Generally it's hard to understand just from the exception what actually happened in a worker. With these changes, one would have to re-run the tests with `get_tracebacks()` put in to get better debugging information.

If the traceback is excessive for a user output, why not mask the `OutOfValues`/`TimedOut` with a high-level exception? 

In fact, ideally, we shouldn't show *any* exceptions to CLI users - any error we can think of should be handled and a user-friendly message shown, with tracebacks put in the log.

Reply:
> The thing that worries me about this approach is that these traceback are a very useful debugging tool

It is useful...in a debugging context - but debugging vs regular running of code are different paradigms. For example, the traceback can be written log file if needed, and then whoever can look at the log file if more context is needed. Alternatively, the calling code can parse the traceback themselves and modify what they want to do with the failures or how they want to represent them. Therefore, client calls can decide what they want to do with the underlying exception traceback instead of being forced to have the stack trace (of a random failure) in the exception message.

Just like regular exceptions.

> If the traceback is excessive for a user output, why not mask the OutOfValues/TimedOut with a high-level exception?

Wrapping the exception is possible but only if something useful can be extracted from the message - what if the entire stack trace is not needed. That would entail something like `e.get_message_without_stack_trace()`

> In fact, ideally, we shouldn't show any exceptions to CLI users - any error we can think of should be handled and a user-friendly message shown, with tracebacks put in the log.

Not sure what you mean by this. 

Currently, the uses of the WorkerPool are for:
    * making arrangements - where the exceptions are caught and ignored - failures are obtained from the WorkerPool itself, not the exception
    * publishing of treasure maps - where the exception is raised from Alice.grant() i.e. called by client code and therefore can be handled (logged or otherwise) as part of writing that client-side code
   
It isn't that we need to show exceptions to users, it is that the exception messages are used for context of what failed. The underlying exception's message is still included in the WorkerPool's exception message, just not the entire stack trace.

Reply:
Further to our conversation - https://github.com/nucypher/nucypher/pull/2772/commits/23140923f504552af289989aa0577dd4a348368c